Storage/src/Connection/Rest.php

```diff
@@ -628,6 +629,53 @@ private function sendWithRetry($resource, $method, $args)
             $args
         );
 
+        $retryIdentifierHash = Uuid::uuid4()->toString();
+
+        $args['onRetryException'] = function (\Exception $e, $currentAttempt, &$arguments) use ($retryIdentifierHash) {
+            // Since we the the last attempt number here, so incrementing it
+            // to get the current attempt count
+            ++$currentAttempt;
+            $this->updateRetryInovationHeaders(
+                $arguments,
+                $retryIdentifierHash,
+                $currentAttempt
+            );
+        };
+
+        $args['onExecutionStart'] = function (&$arguments) use ($retryIdentifierHash) {
+            $this->updateRetryInovationHeaders(
+                $arguments,
+                $retryIdentifierHash
+            );
+        };
+
         return $this->send($resource, $method, $args);
     }
+
+    /**
+     * Updates the x-goog-api-client header value with UUID
+     * and retry count
+     *
+     * @return void
+     */
+    private function updateRetryInovationHeaders(
+        &$arguments,
+        $retryIdentifierHash,
+        $currentAttempt = 0
+    ) {
+        // Fetch existing value(if present) of ['x-goog-api-client']
+        // header value is present and create new required value
+        $request = $arguments[0];
```

Comment:
I am not sure if there are edge cases where the request might be at another index than 0. Although it doesn't look like it.

Probably best to find the index by checking which one implements a `RequestInterface`?

I was planning to do the same in the resumable downloads PR. A small prvt method should do the trick.

Reply:
Yeah, this looks more solid. Implemented the change here [af0e43c](https://github.com/googleapis/google-cloud-php/pull/5741/commits/af0e43c40fb3d06f1eb1649a437cf04305ab765a)